mldsa/ntt.c

```diff
@@ -141,6 +141,41 @@ void ntt(int32_t a[MLDSA_N])
   /* are bounded in magnitude by 9 * MLDSA_Q                 */
 }
 
+/* Reference: Embedded into `invntt_tomont()` in the reference implementation
+ * [@REF] */
+static void mld_invntt_layer(int32_t r[MLDSA_N], unsigned layer)
+__contract__(
+  requires(memory_no_alias(r, sizeof(int32_t) * MLDSA_N))
+  requires(1 <= layer && layer <= 8)
+  requires(array_abs_bound(r, 0, MLDSA_N, MLDSA_Q))
+  assigns(memory_slice(r, sizeof(int32_t) * MLDSA_N))
+  ensures(array_abs_bound(r, 0, MLDSA_N, MLDSA_Q)))
```

Comment:
This post-condition cannot hold as is. 
It is computing `r[j + len] =  r[j]; - r[j + len];`. If inputs are (-q, q), this can be (-2q+1, 2q-1).

Note that in ML-DSA the prime is only 23-bits, so there is a lot of headroom before it overflows the `int32_t`. That is why we need a lot fewer reductions after additions/subtractions than in ML-KEM.

Reply:
Thank you, this helped me get it!